components/scream/src/physics/shoc/shoc_main_impl.hpp

```diff
@@ -285,7 +286,267 @@ void Functions<S,D>::shoc_main_internal(
+  const Int&                   shcol,        // Number of columns
+  const Int&                   nlev,         // Number of levels
+  const Int&                   nlevi,        // Number of levels on interface grid
+  const Int&                   npbl,         // Maximum number of levels in pbl from surface
+  const Int&                   nadv,         // Number of times to loop SHOC
+  const Int&                   num_qtracers, // Number of tracers
+  const Scalar&                dtime,        // SHOC timestep [s]
+  // Input Variables
+  const view_1d<const Scalar>& host_dx,
+  const view_1d<const Scalar>& host_dy,
+  const view_2d<const Spack>& zt_grid,
+  const view_2d<const Spack>& zi_grid,
+  const view_2d<const Spack>& pres,
+  const view_2d<const Spack>& presi,
+  const view_2d<const Spack>& pdel,
+  const view_2d<const Spack>& thv,
+  const view_2d<const Spack>& w_field,
+  const view_1d<const Scalar>& wthl_sfc,
+  const view_1d<const Scalar>& wqw_sfc,
+  const view_1d<const Scalar>& uw_sfc,
+  const view_1d<const Scalar>& vw_sfc,
+  const view_2d<const Spack>& wtracer_sfc,
+  const view_2d<const Spack>& inv_exner,
+  const view_1d<const Scalar>& phis,
+  // Workspace Manager
+  const WorkspaceMgr&      workspace_mgr,
+  // Input/Output Variables
+  const view_2d<Spack>&       host_dse,
+  const view_2d<Spack>&       tke,
+  const view_2d<Spack>&       thetal,
+  const view_2d<Spack>&       qw,
+  const uview_2d<Spack>&      u_wind,
+  const uview_2d<Spack>&      v_wind,
+  const view_2d<Spack>&       wthv_sec,
+  const view_3d<Spack>&       qtracers,
+  const view_2d<Spack>&       tk,
+  const view_2d<Spack>&       shoc_cldfrac,
+  const view_2d<Spack>&       shoc_ql,
+  // Output Variables
+  const view_1d<Scalar>&      pblh,
+  const view_2d<Spack>&       shoc_ql2,
+  // Diagnostic Output Variables
+  const view_2d<Spack>&       shoc_mix,
+  const view_2d<Spack>&       w_sec,
+  const view_2d<Spack>&       thl_sec,
+  const view_2d<Spack>&       qw_sec,
+  const view_2d<Spack>&       qwthl_sec,
+  const view_2d<Spack>&       wthl_sec,
+  const view_2d<Spack>&       wqw_sec,
+  const view_2d<Spack>&       wtke_sec,
+  const view_2d<Spack>&       uw_sec,
+  const view_2d<Spack>&       vw_sec,
+  const view_2d<Spack>&       w3,
+  const view_2d<Spack>&       wqls_sec,
+  const view_2d<Spack>&       brunt,
+  const view_2d<Spack>&       isotropy)
+{
+  using ExeSpace = typename KT::ExeSpace;
 
+  static constexpr Int num_local_scalars = 13;
+  static constexpr Int num_local_ws      = 5;
+  static constexpr Int num_total_local   = num_local_scalars + num_local_ws;
+
+  // Need a second workspace for storing shcol local scalars. Need workspaces
+  // of size schol, not nlev, to store temporary scalars for small kernels
+  const auto policy = ekat::ExeSpaceUtils<ExeSpace>::get_default_team_policy(shcol, 1);
+  ScalarWorkspaceMgr wsm_local(shcol, num_local_scalars, policy, 1.e10); // huge overprovision => no ws sharing
```

Comment:
Since both the workspace size and type are different, I decided to make a new WSM for managing local scalars instead of trying to leverage the existing WSM.